File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-6.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashSet<Character> letters = new HashSet<>();

        int left = 0;
        int result = 0;

        for (int right = 0; right < s.length(); right++) {
            while (letters.contains(s.charAt(right))) {
                letters.remove(s.charAt(left));
                left++;
            }
            letters.add(s.charAt(right));
            result = Math.max(result, letters.size());
        }

        return result;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-6.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        counts = {}
        result = 0

        left = 0
        for right in range(len(s)):
            counts[s[right]] = 1 + counts.get(s[right], 0)

            while (right - left + 1) - max(counts.values()) > k:
                counts[s[left]] -= 1
                left += 1

            result = max(result, right - left + 1)

        return result